snafu-derive/src/lib.rs

```diff
@@ -98,6 +99,76 @@ impl Transformation {
     }
 }
 
+/// SyntaxErrors is a convenience wrapper for a list of syntax errors discovered while parsing
+/// something that derives Snafu.  It makes it easier for developers to add and return syntax
+/// errors while walking through the parse tree.
+struct SyntaxErrors {
+    inner: Vec<syn::Error>,
+}
+
+impl SyntaxErrors {
+    fn new() -> SyntaxErrors {
+        SyntaxErrors { inner: Vec::new() }
+    }
+
+    /// Adds a new syntax error.  The given description will be used in the compile error pointing
+    /// to the given span.  Helper structs are available to format common descriptions, e.g.
+    /// OnlyValidOn and DuplicateAttribute.
+    fn add<D>(&mut self, span: proc_macro2::Span, description: D)
+    where
+        D: fmt::Display,
+    {
+        self.inner.push(syn::Error::new(span, description));
+    }
+
+    /// Consume the SyntaxErrors, returning Ok if there were no syntax errors added, or Err(list)
+    /// if there were syntax errors.
+    fn finish(self) -> MultiSynResult<()> {
+        if self.inner.is_empty() {
+            Ok(())
+        } else {
+            Err(self.inner)
+        }
+    }
+}
+
+/// Helper structure to simplify parameters to SyntaxErrors.add, handling cases where an attribute
+/// was used on an element where it's not valid.
+struct OnlyValidOn {
+    /// The name of the attribute that was misused.
+    attribute: &'static str,
+    /// A description of where that attribute is valid.
+    valid_on: &'static str,
+    /// A description of where the attribute was incorrectly used.
+    not_on: &'static str,
+}
+
+impl fmt::Display for OnlyValidOn {
+    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
```

Comment:
Sadly, `Foo<'_>` isn't valid in older versions of Rust, so we'll have to leave that off for now.

Reply:
Sorry - sure, that's trivial.  I had just copied the `fn fmt` declaration from my current Rust source :)